[PR-39] Improve serialization performance for deeply-nested products.

henryem opened the pull request:
This PR fixes issue #39: "Serialization performance is poor for deeply-nested objects."  I will summarize the issue described in that ticket, then go through this proposed solution.

For objects with nested structure, the algorithm currently implemented in `LowPriorityPBWriterImplicits.prodWriter` performs poorly. For example, if we construct a payload of size K nested inside D layers of `Tuple1` (or single-field case classes), then the write performance is O(D(D+K)) with a large cost on memory allocation and copy. This is because at each layer prodWriter will allocate and copy from a buffer containing the payload plus the headers for the layers beneath. The performance is more complicated with real-world structures, but this example gets the point across, and it is easily verified by a performance benchmark, which I have provided in `PBWriterBenchmark.scala`.

I hope to use PBDirect for serializing some intermediate checkpoint data.  For that real-world data, the write performance improvement due to this fix is around 20-30%, and memory pressure is also greatly reduced.

There are 2 ideas behind this fix:

1. There is a good reason for the existing implementation: the size field for a product comes before the serialized data inside the product, and things must be written sequentially.  So without a way to know the size before serializing the data, it was necessary to precompute the serialized bytes in a separate buffer, then (inside `writeByteArray`) write the size of the buffer, then write the bytes.

  This fix therefore adds a way to compute the size of a product without actually serializing it.  This entails being able to compute the size of any substructure inside the product; therefore every `PBWriter` must implement a new `writtenBytesSize` method.

2. Size computation can be expensive; in fact in the case of strings with high code-points it can be as expensive as serialization itself.  Therefore it's important to avoid redundant size computation through memoization.  So we pass around an `IdentityHashMap[Any, Int]` (aliased to `SizeMap`) to record the sizes of any products encountered in the object graph.

Some notes on the changes:

* Unfortunately, the changes are large in terms of lines of code.  However, the only changed implementation of `writeTo` is in `prodWriter`.  I suggest that you start reviewing from there.
* I've added some unit tests to cover cases that exposed bugs during development of this PR.  Hopefully these are of independent value.
* I've added some performance benchmarks for deeply-nested products.  I would be happy to add more benchmarks if you're worried about the effect of this change on other cases.
* I could not figure out a simple way to handle `contramap` without potentially computing the mapping function twice (once in size computation, and once in serialization).  Hopefully there are no use cases for doing expensive computation inside `contramap`.  If this concerns you, we could think about a way to cache the output of the mapping function, but I wanted to keep this PR manageable.

henryem commented:
The remaining code coverage drop is due to CNil (which wasn't covered before and doesn't seem to need coverage) having an extra line of code.

henryem commented:
@btlines any comment on this?  Note that the reason for the code coverage drop is that the CNil writer has an extra line of code that is not covered, but CNil writing was not covered before (and I can't think of a way to actually cover it).

btlines commented on src/main/scala/pbdirect/PBWriter.scala:
```diff
@@ -1,121 +1,223 @@
+        sizes.get(value)
+      } else {
+        val size = writtenBytesSize(index, value, sizes)
+        sizes.put(value, size)
```
I'm a bit worried here as the size depends on the index but the cached size is only indexed by the value. That's an issue in case of messages containing 2 identical values with different index.

henryem replied:
That's a great point.  Will fix and add a test case.  Hopefully the fix (which I think will involve changing the `sizes` map to a `Map[AnyRef, Map[Int, Int]]`) doesn't impact performance too much.

btlines commented:
Hi @henryem,

Thanks for looking into this, that looks sensible to me.

No worries for the coverage :) still high enough

However I just merged a big PR which conflicts with these changes. Let me know if you need help to get this sorted.

henryem commented:
@btlines I've fixed the merge conflicts and fixed the bug you pointed out (with a unit test), but I am running into another problem: `java.util.IdentityHashMap`, which I need for memoizing the Product sizes, doesn't exist in Scala.js.  I have no experience with Scala.js; any suggestions?

henryem commented:
@btlines I figured out a workaround for the issue with `IdentityHashMap` and also removed my performance benchmark code because it wasn't compatible with Scala 2.13 (unfortunately).  Looks like the build is okay now.

henryem commented:
@btlines I’m ready to merge; any further comments?

btlines commented on shared/src/main/scala/pbdirect/PBWriter.scala:
```diff
@@ -311,110 +534,233 @@ trait PBConsWriter16 extends PBConsWriter8 {
+
+
+//
+//trait LowPriorityPBWriterImplicits {
```
Can you remove this commented section ?

henryem replied:
Thanks for pointing this out.

btlines commented:
LGTM (just one little thing to clean up) and thanks for taking care of all the conflicts.
